src/imitation/density_baselines.py

```diff
@@ -0,0 +1,267 @@
+      ])
+    else:
+      raise ValueError(f"Unknown density type {self.density_type}")
+
+  def __call__(self, obs_b, act_b, next_obs_b, *, steps=None):
+    r"""Compute reward from given (s,a,s') transition batch. This handles
+    *batches* of observations, since it's designed to work with VecEnvs.
+
+    Args:
+      obs_b (np.ndarray): current batch of observations.
+      act_b (np.ndarray): batch of actions that agent took in response to those
+        observations.
+      next_obs_b (np.ndarray): batch of observations encountered after the
+        agent took those actions.
+      steps (np.ndarray): number of time steps elapsed in each environment.
+
+    Returns:
+      rew_array (np.ndarray): array of scalar rewards of the form `r_t(s,a,s')
+        = \log \hat p_t(s,a,s')` (one for each environment), where `\log \hat
+        p` is the underlying density model (and may be independent of s', a, or
+        t, depending on options passed to constructor)."""
+    rew_list = []
+    assert len(obs_b) == len(act_b) and len(obs_b) == len(next_obs_b)
+    for idx, (obs, act, next_obs) in enumerate(zip(obs_b, act_b, next_obs_b)):
+      flat_trans = self._flatten_transition(obs, act, next_obs)
+      scaled_padded_trans = self._scaler.transform(flat_trans[None])
+      if self.is_stationary:
+        rew = self._density_model.score(scaled_padded_trans)
+      else:
+        time = steps[idx]
+        if time >= len(self._density_models):
+          # Can't do anything sensible here yet. Correct solution is to use
+          # hierarchical model in which we first check whether state is
+          # absorbing, then assign either constant score or a score based on
+          # density.
+          raise Exception(
+              f"Time {time} out of range (0, {len(self._density_models)}], and "
+              f"I haven't implemented absorbing states etc. yet")
+        else:
+          time_model = self._density_models[time]
+          rew = time_model.score(scaled_padded_trans)
+      rew_list.append(rew)
+    rew_array = np.asarray(rew_list, dtype='float32')
+    return rew_array
+
+
+class DensityTrainer:
+  def __init__(self,
+               env,
+               rollouts,
+               imitation_trainer,
+               *,
+               standardise_inputs=True,
+               kernel='gaussian',
+               kernel_bandwidth=0.5,
+               density_type=STATE_ACTION_DENSITY,
+               is_stationary=False):
+    r"""Family of simple imitation learning baseline algorithms that apply RL to
+    maximise a rough density estimate of the demonstration trajectories.
+    Specifically, it constructs a non-parametric estimate of `p(s)`, `p(s,s')`,
+    `p_t(s,a)`, etc. (depending on options), then rewards the imitation learner
+    with `r_t(s,a,s')=\log p_t(s,a,s')` (or `\log p(s,s')`, or whatever the
+    user wants the model to condition on).
+
+    Args:
+      env (gym.Env or str): environment to train on.
```

Comment:
Missing `rollouts` arg

Reply:
Could use `TrajectoryList` [named type](https://github.com/HumanCompatibleAI/imitation/blob/a0f2f7b88212851058ba6224b60750fae01d8c4d/src/imitation/util/rollout.py#L19) for doc.